replica/table.cc

```diff
@@ -2837,4 +2838,210 @@ bool table::requires_cleanup(const sstables::sstable_set& set) const {
     }));
 }
 
+unsigned compaction_group::sstable_set::x_log2_compaction_groups(size_t number_of_cgs) {
+    if (number_of_cgs & (number_of_cgs - 1)) {
+        on_internal_error(tlogger, format("Number of compaction groups ({}) is not aligned to power-of-two.", number_of_cgs));
+    }
+    return std::log2(number_of_cgs);
+}
+
+compaction_group::sstable_set::sstable_sets_t
+compaction_group::sstable_set::initialize_sstable_sets(const std::vector<std::unique_ptr<compaction_group>>& cgs) const {
+    if (cgs.empty()) {
+        on_internal_error(tlogger, "Compaction group SSTable set cannot be initialized with a empty list.");
+    }
+    return boost::copy_range<sstable_sets_t>(cgs | boost::adaptors::transformed([] (const std::unique_ptr<compaction_group>& cg) {
+        return std::make_pair(cg->token_range(), cg->make_compound_sstable_set());
```

Comment:
We should filter out empty groups, corresponding to tablets on other nodes or shards.